File: app/src/main/java/com/zxcx/zhizhe/ui/home/HomeFragment.java
package com.zxcx.zhizhe.ui.home;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.zxcx.zhizhe.R;
import com.zxcx.zhizhe.event.GotoHomeRankEvent;
import com.zxcx.zhizhe.mvpBase.BaseFragment;
import com.zxcx.zhizhe.mvpBase.BaseRxJava;
import com.zxcx.zhizhe.mvpBase.IGetPresenter;
import com.zxcx.zhizhe.retrofit.AppClient;
import com.zxcx.zhizhe.retrofit.BaseSubscriber;
import com.zxcx.zhizhe.ui.home.attention.AttentionFragment;
import com.zxcx.zhizhe.ui.home.hot.HotFragment;
import com.zxcx.zhizhe.ui.home.rank.RankFragment;
import com.zxcx.zhizhe.ui.search.search.SearchActivity;
import com.zxcx.zhizhe.ui.search.search.SearchBean;
import com.zxcx.zhizhe.utils.ScreenUtils;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

public class HomeFragment extends BaseFragment implements IGetPresenter<SearchBean> {

    @BindView(R.id.tl_home)
    TabLayout mTlHome;
    Unbinder unbinder;
    @BindView(R.id.app_bar_layout)
    AppBarLayout mAppBarLayout;
    @BindView(R.id.fl_home)
    FrameLayout mFlHome;
    @BindView(R.id.tv_home_search)
    TextView mTvHomeSearch;
    private HotFragment mHotFragment = new HotFragment();
    private AttentionFragment mAttentionFragment = new AttentionFragment();
    private RankFragment mRankFragment = new RankFragment();
    private Fragment mCurrentFragment = new Fragment();

    private String[] titles = new String[]{"推荐", "关注", "榜单"};

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_home, container, false);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        EventBus.getDefault().register(this);
        for (int i = 0; i < titles.length; i++) {
            TabLayout.Tab tab = mTlHome.newTab();
            tab.setCustomView(R.layout.tab_home);
            TextView textView = tab.getCustomView().findViewById(R.id.tv_tab_home);
            textView.setText(titles[i]);
            mTlHome.addTab(tab);
//            tab.setText(titles[i]);
            getSearchHot();
        }
        mTlHome.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                switch (tab.getPosition()) {
                    case 0:
                        switchFragment(mHotFragment);
                        break;
                    case 1:
                        switchFragment(mAttentionFragment);
                        break;
                    case 2:
                        switchFragment(mRankFragment);
                        break;
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });

        ViewGroup.LayoutParams para = mTlHome.getLayoutParams();
        int screenWidth = ScreenUtils.getScreenWidth(); //屏幕宽度
        para.width = screenWidth * 2 / 3;
        mTlHome.setLayoutParams(para);

        initAppBarLayout();
        mTlHome.getTabAt(0).select();
        switchFragment(mHotFragment);
    }

    private void switchFragment(Fragment newFragment) {

        FragmentManager fm = getChildFragmentManager();
        FragmentTransaction transaction = fm.beginTransaction();

        if (newFragment.isAdded()) {
            //.setCustomAnimations(R.anim.fragment_anim_left_in,R.anim.fragment_anim_right_out)
            transaction.hide(mCurrentFragment).show(newFragment).commitAllowingStateLoss();
        } else {
            transaction.hide(mCurrentFragment).add(R.id.fl_home, newFragment).commitAllowingStateLoss();
        }
        mCurrentFragment = newFragment;
    }

    private void initAppBarLayout() {
        mAppBarLayout.addOnOffsetChangedListener((appBarLayout, i) -> {
            mHotFragment.setAppBarLayoutVerticalOffset(i);
            mAttentionFragment.setAppBarLayoutVerticalOffset(i);
            mRankFragment.setMAppBarLayoutVerticalOffset(i);
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        EventBus.getDefault().unregister(this);
        unbinder.unbind();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(GotoHomeRankEvent event) {
        mTlHome.getTabAt(2).select();
    }

    @OnClick(R.id.tv_home_search)
    public void onViewClicked() {
        Intent intent = new Intent(getContext(), SearchActivity.class);
        startActivity(intent);
    }

    public void getSearchHot() {
        mDisposable = AppClient.getAPIService().getSearchHot()
                .compose(BaseRxJava.INSTANCE.handleArrayResult())
                .compose(BaseRxJava.INSTANCE.io_main())
                .subscribeWith(new BaseSubscriber<List<SearchBean>>(this) {
                    @Override
                    public void onNext(List<SearchBean> list) {
                        if (list != null && list.size() > 0) {
                            getDataSuccess(list.get(0));
                        }
                    }
                });
        addSubscription(mDisposable);
    }

    @Override
    public void getDataSuccess(SearchBean bean) {
        mTvHomeSearch.setText(bean.getConent());
    }

    @Override
    public void getDataFail(String msg) {

    }
}

File: app/src/main/java/com/zxcx/zhizhe/ui/home/HomeFragment.kt
package com.zxcx.zhizhe.ui.home

import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.zxcx.zhizhe.R
import com.zxcx.zhizhe.event.HomeClickRefreshEvent
import com.zxcx.zhizhe.event.LogoutEvent
import com.zxcx.zhizhe.mvpBase.BaseFragment
import com.zxcx.zhizhe.ui.home.attention.AttentionFragment
import com.zxcx.zhizhe.ui.home.hot.HotFragment
import com.zxcx.zhizhe.ui.home.rank.RankActivity
import com.zxcx.zhizhe.ui.my.creation.CreationAgreementDialog
import com.zxcx.zhizhe.ui.my.creation.newCreation.CreationEditorActivity
import com.zxcx.zhizhe.ui.my.selectAttention.SelectAttentionActivity
import com.zxcx.zhizhe.ui.my.writer_status_writer
import com.zxcx.zhizhe.ui.search.search.SearchActivity
import com.zxcx.zhizhe.utils.SVTSConstants
import com.zxcx.zhizhe.utils.ScreenUtils
import com.zxcx.zhizhe.utils.SharedPreferencesUtil
import com.zxcx.zhizhe.utils.startActivity
import kotlinx.android.synthetic.main.fragment_home.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

class HomeFragment : BaseFragment() {

    private val mHotFragment = HotFragment()
    private val mAttentionFragment = AttentionFragment()
    private var mCurrentFragment = Fragment()

    private var llRankOffset: Float = 0f

    private val titles = arrayOf("推荐", "关注")

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_home, container, false)
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        EventBus.getDefault().register(this)
        for (i in titles.indices) {
            val tab = tl_home.newTab()
            tab.setCustomView(R.layout.tab_home)
            val textView = tab.customView?.findViewById<TextView>(R.id.tv_tab_home)
            textView?.text = titles[i]
            tl_home.addTab(tab)
            //            tab.setText(titles[i]);
        }
        tl_home.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                when (tab.position) {
                    0 -> {
                        switchFragment(mHotFragment)
                        tv_home_type.text = "推荐"
                    }
                    1 -> {
                        switchFragment(mAttentionFragment)
                        tv_home_type.text = "关注"
                    }
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {}

            override fun onTabReselected(tab: TabLayout.Tab) {

            }
        })

        val para = tl_home.layoutParams
        val screenWidth = ScreenUtils.getScreenWidth() //屏幕宽度
        para.width = screenWidth * 1 / 2
        tl_home.layoutParams = para

        tl_home.getTabAt(0)?.select()
        switchFragment(mHotFragment)

        initAppbarLayout()

        initView()
    }

    private fun initView() {
        tv_home_creation.setOnClickListener {
            if (checkLogin()) {
                when (SharedPreferencesUtil.getInt(SVTSConstants.writerStatus, 0)) {
                    writer_status_writer -> {
                        //创作界面
                        mActivity.startActivity(CreationEditorActivity::class.java,{})
                    }
                    else -> {
                        val dialog = CreationAgreementDialog()
                        dialog.show(mActivity.fragmentManager, "")
                    }
                }
            }
        }
        tv_home_rank.setOnClickListener {
            mActivity.startActivity(RankActivity::class.java,{})
        }
        view_home_search.setOnClickListener {
            mActivity.startActivity(SearchActivity::class.java,{})
        }
        iv_home_search.setOnClickListener {
            mActivity.startActivity(SearchActivity::class.java,{})
        }
        iv_attention_add.setOnClickListener {
            mActivity.startActivity(SelectAttentionActivity::class.java,{})
        }
    }

    private fun initAppbarLayout() {
        app_bar_layout.addOnOffsetChangedListener { appBarLayout, verticalOffset ->
            if (llRankOffset == 0f){
                llRankOffset = ll_home_rank.top.toFloat()
            }
            val total = appBarLayout.totalScrollRange * 1.0f
            //计算出滑动百分比
            val p = Math.abs(verticalOffset) / total
            tv_home_title.alpha = 1 - p
            view_home_search.alpha = 1 - p
            ll_home_rank.alpha = 1 - p
            tv_home_type.alpha = 1 - p

            if (Math.abs(verticalOffset) > llRankOffset){
                ll_home_rank.alpha = 0f
                ll_home_tab.visibility = View.VISIBLE
                ll_home_tab.alpha = if (p == 1f) p else p/2
                if (mCurrentFragment == mAttentionFragment){
                    iv_attention_add.visibility = View.VISIBLE
                }else{
                    iv_attention_add.visibility = View.GONE
                }
            }else{
                ll_home_tab.visibility = View.INVISIBLE
            }
        }
    }

    private fun switchFragment(newFragment: Fragment) {

        val fm = childFragmentManager
        val transaction = fm.beginTransaction()

        if (newFragment.isAdded) {
            //.setCustomAnimations(R.anim.fragment_anim_left_in,R.anim.fragment_anim_right_out)
            transaction.hide(mCurrentFragment).show(newFragment).commitAllowingStateLoss()
        } else {
            transaction.hide(mCurrentFragment).add(R.id.fl_home, newFragment).commitAllowingStateLoss()
        }
        mCurrentFragment = newFragment
    }

    override fun onDestroyView() {
        super.onDestroyView()
        EventBus.getDefault().unregister(this)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: HomeClickRefreshEvent) {
        app_bar_layout.setExpanded(true)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: LogoutEvent) {
        app_bar_layout.setExpanded(true)
    }
}
